src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,557 @@
+   * partitioned and then sorted per partition. This enables the join to perform only linear reads and writes between
+   * NUMA clusters, avoiding the latency assiociated with these operations.
+   * The implementation of NUMA awareness hinges on explicitly allocating data on NUMA nodes using pmr_allocators and
+   * scheduling operations to always run on the same node as the data.
+**/
+
+namespace opossum {
+JoinMPSM::JoinMPSM(const std::shared_ptr<const AbstractOperator> left,
+                   const std::shared_ptr<const AbstractOperator> right, const JoinMode mode,
+                   const std::pair<ColumnID, ColumnID>& column_ids, const PredicateCondition op)
+    : AbstractJoinOperator(OperatorType::JoinMPSM, left, right, mode, column_ids, op) {
+  // Validate the parameters
+  DebugAssert(mode != JoinMode::Cross, "This operator does not support cross joins.");
+  DebugAssert(left != nullptr, "The left input operator is null.");
+  DebugAssert(right != nullptr, "The right input operator is null.");
+  DebugAssert(op == PredicateCondition::Equals, "Only Equi joins are supported by MPSM join.");
+}
+
+std::shared_ptr<const Table> JoinMPSM::_on_execute() {
+  // Check column types
+  const auto& left_column_type = input_table_left()->column_data_type(_column_ids.first);
+  DebugAssert(left_column_type == input_table_right()->column_data_type(_column_ids.second),
+              "Left and right column types do not match. The mpsm join requires matching column types");
+
+  // Create implementation to compute the join result
+  _impl = make_unique_by_data_type<AbstractJoinOperatorImpl, JoinMPSMImpl>(
+      left_column_type, *this, _column_ids.first, _column_ids.second, _predicate_condition, _mode);
+
+  return _impl->_on_execute();
+}
+
+void JoinMPSM::_on_cleanup() { _impl.reset(); }
+
+std::shared_ptr<AbstractOperator> JoinMPSM::_on_recreate(
+    const std::vector<AllParameterVariant>& args, const std::shared_ptr<AbstractOperator>& recreated_input_left,
+    const std::shared_ptr<AbstractOperator>& recreated_input_right) const {
+  return std::make_shared<JoinMPSM>(recreated_input_left, recreated_input_right, _mode, _column_ids,
+                                    _predicate_condition);
+}
+
+const std::string JoinMPSM::name() const { return "Join MPSM"; }
+
+template <typename T>
+class JoinMPSM::JoinMPSMImpl : public AbstractJoinOperatorImpl {
+ public:
+  JoinMPSMImpl<T>(JoinMPSM& mpsm_join, ColumnID left_column_id, ColumnID right_column_id, const PredicateCondition op,
+                  JoinMode mode)
+      : _mpsm_join{mpsm_join},
+        _left_column_id{left_column_id},
+        _right_column_id{right_column_id},
+        _op{op},
+        _mode{mode} {
+    _topology = _get_or_create_numa_topology();
+    _cluster_count = _determine_number_of_clusters();
+    _output_pos_lists_left.resize(_cluster_count);
+    _output_pos_lists_right.resize(_cluster_count);
+    for (size_t cluster = 0; cluster < _cluster_count; ++cluster) {
```

Comment:
if we have a typedef for `ClusterID`, then this should also be `ClusterID` and not `size_t`

Reply:
this applies to quite a few `size_t`s in the following code